posts/ast_primer.md

```diff
@@ -0,0 +1,405 @@
+# A Primer on SQLGLot's Abstract Syntax Tree
+
+SQLGlot is a powerful tool for analyzing and transforming SQL, but the learning curve can be intimidating.
+ 
+This post is intended to familiarize newbies with SQLGlot's abstract syntax trees, how to traverse them, and how to mutate them.
+
+## The tree
+
+SQLGlot parses SQL into an abstract syntax tree (AST).
+```python
+from sqlglot import parse_one
+
+ast = parse_one("SELECT a FROM (SELECT a FROM x) AS x")
+```
+
+An AST is a data structure that represents a SQL statement. The best way to glean the structure of a particular AST is python's builtin `repr` function:
+```python
+repr(ast)
+
+# (SELECT expressions: 
+#   (COLUMN this: 
+#     (IDENTIFIER this: a, quoted: False)), from: 
+#   (FROM this: 
+#     (SUBQUERY this: 
+#       (SELECT expressions: 
+#         (COLUMN this: 
+#           (IDENTIFIER this: a, quoted: False)), from: 
+#         (FROM this: 
+#           (TABLE this: 
+#             (IDENTIFIER this: x, quoted: False)))), alias: 
+#       (TABLEALIAS this: 
+#         (IDENTIFIER this: x, quoted: False)))))
+```
+
+This is a textual representation of the internal data structure. Here's a breakdown of some its components:
+```
+Expression type    child key
+  |               /
+(SELECT expressions:
+  (COLUMN this:  ---------------------------------- COLUMN is a child node of SELECT
+    (IDENTIFIER this: a, quoted: False)), from:  -- "from:" is another child key of SELECT
+  (FROM this: ------------------------------------- FROM is also a child node of SELECT
+  ...
+```
+
+## Nodes of the tree
+
+The nodes in this tree are instances of `sqlglot.Expression`. Nodes reference their children in `args` and their parent in `parent`:
+```python
+ast.args
+# {
+#    "expressions": [(COLUMN this: ...)],
+#    "from": (FROM this: ...),
+#    ...
+# }
+
+ast.args["expressions"][0]
+# (COLUMN this: ...)
+
+ast.args["expressions"][0].args["this"]
+# (IDENTIFIER this: ...)
+
+ast.args["from"]
+# (FROM this: ...)
+
+assert ast.args["expressions"][0].args["this"].parent.parent is ast
+```
+
+Children can either be:
+1. An Expression instance
+2. A list of Expression instances
+3. Another Python object, such as str or bool. This will always be a leaf node in the tree.
+
+Navigating this tree requires an understanding of the different Expression types. The best way to browse Expression types is directly in the code at [expressions.py](../sqlglot/expressions.py). Let's look at a simplified version of one Expression type:
+```python
+class Column(Expression):
+    arg_types = {
+      "this": True, 
+      "table": False,
+      ...
+    }
+```
+
+`Column` subclasses `Expression`.
+
+`arg_types` is a class attribute that specifies the possible children. The `args` keys of an Expression instance correspond to the `arg_types` keys of its class. The values of the `arg_types` dict are `True` if the key is required, meaning it is guaranteed to be present in `args`.
```

Comment:
> meaning it is guaranteed to be present in `args`.

There is a gotcha / small detail here that you can bypass that guarantee and still create a "partial" AST if you want, for example using the `IGNORE` error flag. Do you think there's value in hinting that here or in a footnote?

Reply:
ah, good to know. I'll use softer language.